[WIP] Mage Redesign/Rework

Sharrq opened the pull request:
Over years of iteration and work on WoWAnalyzer and the Mage specs, keeping things updated has become more difficult and more complicated which has made things very time consuming to update. My goal with this rework is to make things a little more generalized, add more templates and helpers and generally make things easier to understand and manage overall, particularly for someone that does not have a lot of programming experience (like myself).

I kept the rework contained to the mage class and to Arcane only for now while I get feedback and iterate on things. This rework was largely done by copilot so there are almost certainly things that can be improved or iterated on further. 

As stated, I am not a programmer so I have no way of knowing if any of the things copilot did are good or not. My goal here is to do the fundamental rework ahead of Midnight and then update things afterwards. There are certainly things that are not evaluating properly or may be giving bad advice spec wise, but the goal here is to rebuild and rework things and then update the logic afterwards.

emallson reviewed (commented):
this is a very large PR, so it is going to take a bit to get to/through. i have skimmed it, and there are definitely some interesting ideas here. i am pretty interested in things that streamline support, and will be taking a look at that in particular.

that said, i did want to comment early about mostly-llm PRs and about this PR in particular:

- even if an LLM wrote the code, it is still your responsibility as the person submitting the PR to know what is being submitted and (at least *mostly*) understand it. i can't ask the LLM *why* something exists or what the *intent* was, but I can ask you those questions

-  > As stated, I am not a programmer so I have no way of knowing if any of the things copilot did are good or not.

You've been working on this project for a long time. You're definitely qualified to have an opinion on whether something is good, even if your day job isn't software dev.

- two of the biggest problems i've seen in reviewing LLM code are (1) it is *awful* at handling spell analysis. don't trust it at all for that, and (2) it has a tendency to generate a ton of code, which is inflated by *duplicate implementations*. i saw a couple of those already when skimming through this PR so i have no doubt that is inflating the line count here.

all that said i'm already a bit behind on wowa reviews so i don't expect a detailed review of this right away

Sharrq commented:
@emallson 

What I meant when I said I'm not a programmer and don't know if this is good, is that I don't know if the methods it is taking are harmful to performance or general things that would be problematic long-term. I have been evaluating it as I have been going, and correcting it when it goes awry, but the best example I can give is when I previously moved everything to use Event History and then tanked the performance of fire mage as a result ... It looked good to me, but I had no idea what the performance impact was.

Also yes, Copilot duplicated a LOT of things. Many many many of them I have caught and fixed already but I am sure there are some that I missed.

I have not been having Copilot do any of the spec analysis .. it was using the existing analysis as is (and some things I noticed issues and fixed those myself), but I largely used Copilot here to redesign things structurally as that is what this PR is mostly focused on. I know the spec analysis is wrong because it was wrong before I started and might be a little bit more wrong now .... but my goal with this was not to bring Arcane up to 11.2, it was to redesign how everything was built ahead of Midnight.

And I knew ahead of time that this was going to be massive and take time to review. It is part of the reason I kept everything isolated to Mage so that if anything was missed it will only affect the specs that I maintain. Also this is largely something I am working on with the intention of making things easier for Midnight, so if it takes a bit longer that's fine. 

I am still reviewing things and making adjustments as I review it myself, but given the size of this I didn't want to wait until it was 100% to start getting feedback. I think I have it 85-90% there, and am in a good place to do some fine tuning based on feedback and such.

emallson commented:
> What I meant when I said I'm not a programmer and don't know if this is good, is that I don't know if the methods it is taking are harmful to performance or general things that would be problematic long-term. [...]

that makes sense 👍 it's fair to acknowledge a knowledge gap, but don't sell yourself short either!

> I am still reviewing things and making adjustments as I review it myself, but given the size of this I didn't want to wait until it was 100% to start getting feedback. I think I have it 85-90% there, and am in a good place to do some fine tuning based on feedback and such.

👍 i saw the convo in discord so i figured (and this IS still a draft PR) but yea, it is also very large lol

> I have not been having Copilot do any of the spec analysis .. it was using the existing analysis as is (and some things I noticed issues and fixed those myself), but I largely used Copilot here to redesign things structurally as that is what this PR is mostly focused on. I know the spec analysis is wrong because it was wrong before I started and might be a little bit more wrong now .... but my goal with this was not to bring Arcane up to 11.2, it was to redesign how everything was built ahead of Midnight.

with midnight on the horizon i think this is a good approach. sinking a ton of time into 11.2 rework stuff when midnight has massive mage changes seems...unwise

Sharrq commented:
I deleted some modules belonging to spells and talents removed in Midnight. This was mostly to cut down on the number of files involved in the rework for now and make it somewhat vaguely easier to review.

I believe I am done making changes on this PR and will wait for any feedback items. Planning to get this added before I continue reworks for fire next using this PR as a template.

Did a little cleanup across all of these items, removed some duplicate functions, cleaned up the documentation a bit, and moved and reworked the cast timeline pieces and moved them back into Mage from Core

emallson commented on src/analysis/retail/mage/shared/helpers/guideHelpers.tsx:
```diff
@@ -0,0 +1,131 @@
+ * @param isGreaterThan True if higher values are better, false if lower values are better
+ * @returns QualitativePerformance level
+ */
+export function evaluatePerformance(
```
i believe this is a duplicate of `evaluateQualitativePerformanceByThreshold` (which is quite a lengthy name and one that i think a lot of people miss)

emallson commented on src/analysis/retail/mage/shared/helpers/guideHelpers.tsx:
```diff
@@ -0,0 +1,131 @@
+export function evaluateBoolean(condition: boolean, invertLogic = false): QualitativePerformance {
+  const result = invertLogic ? !condition : condition;
+  return result ? QualitativePerformance.Good : QualitativePerformance.Fail;
+}
```
this kind of logic is generally too small to move into a method like this. there are exceptions (sometimes you're writing a function that takes a list of operations and having an `evaluateBoolean` method like this could be useful), but generally wrapping ternary operators like this does more harm than good.

emallson commented on src/analysis/retail/mage/shared/helpers/guideHelpers.tsx:
```diff
@@ -0,0 +1,131 @@
+    getResult: (cast: unknown, evaluatedData: BoxRowEntry) => boolean;
+    getDetails: (cast: unknown) => string;
+  }>;
+}): ExpandableConfig {
```
maybe vestigial, maybe just an LLM helper, but the input and output types here are the same.

emallson commented on src/analysis/retail/mage/shared/components/GeneralizedChart.tsx:
```diff
@@ -0,0 +1,340 @@
+                          : ann.color || this.getAnnotationColor(ann.type),
+                  }}
+                />
+              </div>
```
this is backward from the series legend (the color is *after* for annotations, but *before* for series)

<img width="1294" height="410" alt="image" src="https://github.com/user-attachments/assets/ac884cab-a775-436e-b374-5204368a6791" />

emallson commented on src/analysis/retail/mage/shared/components/GeneralizedChart.tsx:
```diff
@@ -0,0 +1,340 @@
+  endTime: number;
+}
+
+class GeneralizedChart extends React.PureComponent<Props> {
```
this should be a function component instead of a class component. the private methods can simply be un-exported functions in the file (in each case they do not use `this`, and so are effectively static)

emallson commented on src/analysis/retail/mage/shared/helpers/castLinkHelpers.ts:
```diff
@@ -0,0 +1,269 @@
+
+const CAST_BUFFER_MS = 75;
+
+export const EventRelations = {
```
i think this is an LLM-ism but this basically just duplicates `EventType` for no reason

emallson commented on src/analysis/retail/mage/shared/builders/StatisticBuilder.tsx:
```diff
@@ -0,0 +1,335 @@
+   * @param getValue Function to extract the value from data object
+   * @returns This builder for method chaining
+   */
+  column(headerOrConfig: ReactNode | ColumnConfig<T>, getValue?: (data: T) => ReactNode): this {
```
no need for this to be either a config, or a pair of params. pick one.

emallson commented on src/analysis/retail/mage/shared/helpers/guideHelpers.tsx:
```diff
@@ -0,0 +1,131 @@
+ * @param timestamp Timestamp of the event
+ * @returns JSX tooltip element
+ */
+export function generateGuideTooltip(
```
this should be a proper react component, not a function-returning-jsx

emallson commented:
(there's also a bug in your most recent commit where moving files didn't move all references to them. i have a fix locally that i'll push later---its on another machine)

Sharrq commented:
I read through this briefly but will go through it in a bit more detail after work. As for the builders, I had asked copilot to give me some ideas to simplify things so I could pick one I liked, and thought the builders stuff looked clean and easy so I picked it. And then went all in on it because I wanted consistency so it wasn't like guide is using this format, charts use this, statistics are this, etc. But I did start to have some second thoughts on it towards the end. I might reevaluate options. 

And yes I agree that while moving from suggestions to checklist to guide has added a lot of improvements, it has also made things drastically more complicated and it's probably the biggest reason why I've been struggling to keep up with the mage specs. It took me like a weekend to add the initial support for frost mage back in legion when I started and now I feel like to do the same would take weeks or months. That is largely what I am trying to solve here .... 

There are also a lot of things that I really wanted to add to the guide but figuring out how to add it was a nightmare. Things like cast sequences and timelines to show in the guide for important burst windows that are really only used in the apl stuff, massive components that can be very difficult to reverse engineer and adapt, especially for someone with very little or no programming experience. These were not an issue when I started because everything was simpler but now that it's so complex it feels like I can't really do what I want to without spending hours/days trying to get one little chart or drop down to show what I want or whatever. 

As for some of the stuff with single usages. My plan was to adapt fire and frost to this as well, which will help with that a little bit but I'll take a look at them and see if I went too far on some of the helper stuff

Thanks!

Sharrq commented:
@emallson Looking at options for replacing the builders, it sounded like this was closer to what you were thinking as an example?

Guide
- The same condition logic with failConditions, perfectConditions, etc.
```
get guideSubsection(): JSX.Element {
    const explanation = (
      <>
        <b><SpellLink spell={SPELLS.ARCANE_BARRAGE} /></b> is your charge spender...
        {/* ... rest of explanation ... */}
      </>
    );

    return (
      <MageGuideSection 
        spell={SPELLS.ARCANE_BARRAGE}
        explanation={explanation}
      >
        <CastBreakdown data={this.arcaneBarrageData} />
      </MageGuideSection>
    );
  }
  ```
  
  Charts
  ```
  get guideSubsection(): JSX.Element {
    const explanation = (/* ... same explanation ... */);

    return (
      <MageGuideSection 
        spell={TALENTS.EVOCATION_TALENT}
        title="Mana Management"
        explanation={explanation}
        verticalLayout
      >
        <ManaChartComponent
          manaUpdates={this.manaUpdates}
          startTime={this.owner.fight.start_time}
          endTime={this.owner.fight.end_time}
          annotations={[
            { 
              events: this.arcaneSurgeCasts, 
              type: 'cast', 
              color: SPELL_COLORS.ARCANE_SURGE 
            },
            { 
              events: this.evocationCasts, 
              type: 'cast', 
              color: SPELL_COLORS.EVOCATION 
            },
          ]}
          lowResourceWarning={{ threshold: 0.1, label: 'Low Mana' }}
          showBossHealth
          reportCode={this.owner.report.code}
        />
      </MageGuideSection>
    );
  }
  ```
  
  Statistics
  ```
  statistic() {
    const dropdown = (
      <StatisticDropdownTable
        headers={['Stacks', 'Haste', 'Time (s)', 'Time (%)']}
        rows={this.timeAtStackCount.map((time, stacks) => ([
          stacks,
          `${formatPercentage(stacks * ARCANE_TEMPO_HASTE_PER_STACK, 0)}%`,
          formatDuration(time),
          `${formatPercentage(time / this.owner.fightDuration)}%`,
        ]))}
      />
    );

    return (
      <MageStatistic
        spell={TALENTS.ARCANE_TEMPO_TALENT}
        position={STATISTIC_ORDER.CORE(7)}
        value={this.averageHaste}
        label="average haste gained"
        format="percentage"
        icon={<HasteIcon />}
        dropdown={dropdown}
      />
    );
    ```

Sharrq commented:
Going to flip this back to WIP. I am reworking things to move away from the builders and still want to iterate on some things a bit.

Sharrq commented:
**Stuff I changed**
- I removed the builders entirely and moved towards creating components that serve a similar purpose.
- I did narrow the charts to just the mana chart since that is the only thing I am using currently
- I went through the helpers files and removed a lot that was offloaded to the new components
- I removed some helpers that were duplicates of core stuff

**Stuff I'm still working on**
- I am going to look at the chart stuff a little more and make sure everything outdated/deprecated got stripped out, I think I still have a little bit to disconnect from GeneralizedChart
- I still want to iterate on Statistics a bit because I like what you pointed out and I don't really like what it did in my most recent commits. I am also open to doing that Statistic overhaul in core, but probably after this PR is finished. So ill probably model it out inside Mage and then follow up with a different PR to move it into core and overhaul it on a larger level.
- Now that I have simplified the EventRelations stuff a bit, I am going to simplify this a bit more. My overall goal with this was to reduce some bloat and confusion by allowing you to specify multiple links from a single source. I have noticed a common issue I have run into is that I lose track of what all I have linked to each spell because each link is its own block and they are often added at different times so they are not together. I wanted to be able to say like "Here is Arcane Barrage's cast event, and here is everything I want you to link to it" so they could all be bundled together. I think I can accomplish that while simplifying it a little bit

Sharrq commented:
Ok, I have reworked my rework a bit after some of the feedback, pulling back on some things and changing directions on some others. 

- I reworked the Statistics, sticking with a similar direction. My goal here is to do this rework on the arcane modules and expand that to an overall rework in a later PR
- I trimmed castLinkHelpers down to a single multipurpose "link" function where you specify the relation instead of having one for every event type and also moved Event relations to just use EventType instead for the existing types and just a const with some custom types for custom stuff where there isn't an existing type.
- I eliminated a lot of the helpers and eliminated all of the builders
- I removed MageAnalyzer and moved things back to Analyzer. There was only one remaining function in MageAnalyzer that seemed useful, and I added it to the core EventHistory instead.
- I eliminated GeneralizedChart and just focused on the ManaChart instead. ManaChart is essentially just ManaLevelChart with some minor improvements, a simplified method of interacting with it in code, and the ability to add annotations to it easily ... I think this could easily replace the core ManaLevelGraph as without the annotations it just does exactly what ManaLevelChart does .. but I of course don't want to do that across all the healer specs without discussing it, and certainly not in this PR.
- I started playing with some new Guide visualization elements including a cast timeline segment (which is something I had been trying to figure out how to add for a while), as well as a couple "card" visualizations that can better show individual details about a spell or summarize various overview type items and allow drilling down.
- I converted my guide rework towards components instead. Similar to Statistics, I think I could do a larger core rework involving building out various premade components that accomplish various common visualizations instead of people needing to build them repeatedly in each guide. Could create some additional visualizations as well alongside this (all in a separate PR). I agree that the majority of the guides all use the same visualizations (Performance Box Rows and bars) because they are the easiest to implement. But I also believe that if there were more premade visualizations that were easy to interact with, then there would be more diversity in visual elements across guides.

Below are screenshots of the new guide visualizations I added. For the overview/detail ones, the info in it is largely placeholder and it isn't perfect, but just wanted to explore some options for showing info in different ways

Cast Overview

<img width="728" height="409" alt="Screenshot 2025-10-18 12 36 05 AM" src="https://github.com/user-attachments/assets/281147ec-8326-489e-a0da-b0ce1272feb2" />

Cast Overview - if you click on one of the performance box rows
<img width="728" height="762" alt="Screenshot 2025-10-18 12 36 26 AM" src="https://github.com/user-attachments/assets/6e8bdcc7-4eba-4ab0-88a5-7e342c1eccdb" />

Cast Detail
<img width="728" height="424" alt="Screenshot 2025-10-18 12 35 48 AM" src="https://github.com/user-attachments/assets/2dcbf773-ae7e-48a4-ab75-7b6303659c6e" />

Cast Timeline
<img width="728" height="270" alt="Screenshot 2025-10-18 12 35 30 AM" src="https://github.com/user-attachments/assets/9f414474-8f91-4c0a-8bf5-d0376c0fd040" />

Sharrq commented:
I have other things I want to do and other ideas I have, but I am going to stop myself here aside from any additional feedback.

emallson commented on src/parser/shared/modules/EventHistory.ts:
```diff
@@ -229,6 +229,43 @@ class EventHistory extends Module {
+   * @param spell optional spell filter(s)
+   * @returns array of events matching the criteria within the time window
+   */
+  getEventsInTimeWindow<ET extends EventType>(
```
the runtime of this is going to be pretty bad i think. probably better to just make `getEvents` understand an array of event types, eliminate this wrapper method, and use `getEvents` directly to retrieve events of multiple types. that would also eliminate the extra sort.

emallson commented on src/analysis/retail/mage/shared/components/statistics/MageStatistic.tsx:
```diff
@@ -111,7 +58,207 @@ export default function MageStatistic({
+/**
+ * Displays custom content.
+ */
+MageStatistic.Value = function Value({ children }: { children: ReactNode }) {
```
better name: `MageStatistic.Custom` or `CustomValue`. `Value` is very vague.

emallson commented on src/analysis/retail/mage/shared/components/guide/ManaChart.tsx:
```diff
@@ -317,10 +338,128 @@ export default function ManaChart({
+        setLoading(true);
+      }
+      try {
+        const fetchWcl = (await import('common/fetchWclApi')).default;
```
this doesn't need to be an `async` import. `fetchWclApi` is part of pretty much every bundle

emallson commented on src/analysis/retail/mage/arcane/guide/PresenceOfMind.tsx:
```diff
@@ -2,108 +2,109 @@ import SPELLS from 'common/SPELLS';
-            },
-          ],
+    // Fail conditions
+    if (Boolean(ST) && !touchAtEnd) {
```
these `Boolean(...)` calls aren't needed

emallson commented on src/analysis/retail/mage/shared/components/guide/CastDetails.tsx:
```diff
@@ -0,0 +1,484 @@
+
+  &:hover {
+    background: rgba(0, 0, 0, 0.4);
+    transform: translateX(2px);
```
this `translate` is very distracting. typically, you'd use something like this to indicate an interactive element---not an entire container. the container itself isn't interactive (as far as I can tell), so the `&:hover` block should likely be cut entirely

emallson commented on src/analysis/retail/mage/shared/helpers/castLinkHelpers.ts:
```diff
@@ -0,0 +1,121 @@
+ * @returns A LinkConfig object
+ *
+ */
+export function link(relation: EventType | string, overrides?: Partial<LinkConfig>): LinkConfig {
```
perhaps overkill for this PR, but you can type this using function overload types ([docs](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)) like:

```tsx
function link(relation: EventType, overrides?: Omit<LinkConfig, 'type' | 'relation'>): LinkConfig;
function link(relation: string, overrides: Omit<LinkConfig, 'relation'>): LinkConfig;
export function link(relation: EventType | string, overrides?: Partial<LinkConfig>): LinkConfig {
 // actual method implementation
}
```

(even with types, the check/throw is fine)

emallson commented on src/analysis/retail/mage/shared/components/guide/CastDetails.tsx:
```diff
@@ -0,0 +1,484 @@
+
+export interface CastDetailsProps {
+  casts: CastEntry[];
+  formatTimestamp: (timestamp: number) => string;
```
rather than this `formatTimestamp` prop, use the `useFight` hook to retrieve the fight (which will let you get the `startTime`) and then use `formatDuration` as normal.

that will allow you to eliminate the `this.owner.formatTimestamp.bind(...)` calls in analyzers

emallson replied:
(actually this one should also get done, unless `useFight` doesn't work within this component for some godforsaken reason)

emallson commented on src/analysis/retail/mage/shared/components/guide/CastDetails.tsx:
```diff
@@ -0,0 +1,484 @@
+    setCurrentIndex(0);
+  };
+
+  const performanceColors = {
+    [QualitativePerformance.Perfect]: '#1eff00',
+    [QualitativePerformance.Good]: '#fab700',
+    [QualitativePerformance.Ok]: '#ff8800',
+    [QualitativePerformance.Fail]: '#ff0000',
+  };
+
+  const performanceLabels = {
+    [QualitativePerformance.Perfect]: 'Perfect',
+    [QualitativePerformance.Good]: 'Good',
+    [QualitativePerformance.Ok]: 'OK',
+    [QualitativePerformance.Fail]: 'Failed',
+  };
```
we really need to standardize these, sigh.

not in this PR, though

emallson commented on src/analysis/retail/mage/shared/components/guide/NoCastsMessage.tsx:
```diff
@@ -0,0 +1,20 @@
+ * Displays message when no casts of a talented spell are recorded.
+ * @param spell - The spell that had no casts
+ */
+export const NoCastsMessage = ({ spell }: NoCastsMessageProps) => (
```
this could be used within `CastsSummary` to provide a good default when no casts are present, instead of the ternary that is currently used.

emallson commented on src/analysis/retail/mage/shared/components/guide/CastOverview.tsx:
```diff
@@ -0,0 +1,328 @@
+            );
+
+            return stat.tooltip ? (
+              <TooltipElement key={index} content={stat.tooltip}>
```
nitpick: this should use `Tooltip` instead of `TooltipElement` because of the dashed underline not playing nicely with the display. (it looks *fine* i guess, but it doesn't read like "this has a tooltip")

emallson commented on src/analysis/retail/mage/shared/components/guide/CastOverview.tsx:
```diff
@@ -0,0 +1,328 @@
+              );
+
+              return detail.tooltip ? (
+                <TooltipElement key={index} content={detail.tooltip}>
```
same nit here about `TooltipElement` vs `Tooltip` (yes, the names aren't good)

emallson commented:
oh also, prettier & e2e need to pass, unless the e2e failures are unrelated.

emallson commented:
couple of other notes on the ui elements:

- i like being able to filter to good/bad/perfect casts. we should do that more, it's useful to have access to everything but quickly filter to fails
- timelines for cooldown windows are nice 👍 might be nice to put the timelines *in* the cooldown dropbox instead of having the pager at the bottom. if that is *mostly* just deleting code (for the pager), might be worth doing in this PR. if it requires significant changes to the cooldown breakdown code, leave it for later.

Sharrq commented:
> for my sanity, please don't add any more new features / refactors to this PR. lets get this one cleaned up and merged into the `midnight` branch, and then new things can be added in follow-ups. (i know you already said you're done adding new things, but i really wanted to emphasize it lol)

Yea, sorry. I didnt want to have you go through everything twice, but moving away from the builders seemed easier to do now than to completely redo it again in another pr (or several PRs). I will work through the comments, but my plan is to clean this up and push it and continue to build in other directions, adjusting things as i go but i think this is a good foundation to work off of

emallson commented:
> Yea, sorry. I didnt want to have you go through everything twice, but moving away from the builders seemed easier to do now than to completely redo it again in another pr (or several PRs).

yea, that's totally fine! i think it was a good decision.

> I will work through the comments, but my plan is to clean this up and push it and continue to build in other directions, adjusting things as i go but i think this is a good foundation to work off of

sounds good 👍

Sharrq commented:
I made the requested changes. i think the e2e errors previously were something timing out iirc. 

Ill have to mess with prettier tomorrow. I thought it ran that automatically when you commit stuff previously, i dont think i have ever had to run it manually.

emallson reviewed (approved):
thanks! once prettier changes are in, you've got a green light to merge

Sharrq commented:
Prettier always used to run on my commits, and im not sure why it doesnt anymore. Do i need to do something locally to get it to do that?

Sharrq commented:
I think i fixed it. didnt have prttier installed in npm. apparently ArcaneSurge was the only file that it was complaining about and i made a trivial change to it and it ran on it
